### PR TITLE
Fixed #23757 -- Added 3D introspection support to Spatialite backend

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

**Comment by timgraham** [comment]
Claude, code looks fine, but I think there were still unaddressed test failures last time I looked at this. Should this be documented?

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
buildbot, test on lucid.

**Comment by timgraham** [comment]
buildbot, test on lucid.

**Comment by timgraham** [comment]
The `Geo3DTest` tests don't work on the version of spatialite on 12.04 - maybe the features need to be a function of the spatial version?

**Comment by claudep** [comment]
I suspect this patch is incompatible with GEOS 3.2. Could you check if Lucid is using GEOS 3.2?

**Comment by timgraham** [comment]
3.1 according to https://launchpad.net/ubuntu/lucid/+source/geos

**Comment by claudep** [comment]
And we don't support 3.1 any longer...
https://docs.djangoproject.com/en/1.8/ref/contrib/gis/install/geolibs/

**Comment by timgraham** [comment]
Oops, 12.04 is called precise, not lucid. So we actually have 3.2.2 there.

**Comment by timgraham** [comment]
buildbot, test on precise.

**Comment by timgraham** [comment]
buildbot, test on precise.

File: django/contrib/gis/db/backends/spatialite/features.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 from django.contrib.gis.db.backends.base.features import BaseSpatialFeatures
-from django.db.backends.sqlite3.features import \
-    DatabaseFeatures as SQLiteDatabaseFeatures
+from django.contrib.gis.geos import geos_version
+from django.db.backends.sqlite3.features import (
+    DatabaseFeatures as SQLiteDatabaseFeatures,
```
Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/features.py]: I'm afraid isort insists on the backslash version

File: django/contrib/gis/db/backends/spatialite/introspection.py [hunk 2]
```diff
@@ -41,7 +41,13 @@ def get_geometry_type(self, table_name, geo_col):
 
             # OGRGeomType does not require GDAL and makes it easy to convert
             # from OGC geom type name to Django field.
-            field_type = OGRGeomType(row[2]).django
+            ogr_type = row[2]
+            if isinstance(ogr_type, six.integer_types) and ogr_type > 1000:
+                # Spatialite versions >= 4 use the new SFSQL 1.2 offsets
+                # 1000 (Z), 2000 (M) and 3000 (ZM) to indicate the presence of
```
Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/introspection.py]: comma before "and"

File: django/contrib/gis/db/backends/spatialite/introspection.py [hunk 3]
```diff
@@ -41,7 +41,13 @@ def get_geometry_type(self, table_name, geo_col):
...
+                # higher dimensional coordinates (M not yet supported by Django)
```
Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/introspection.py]: add period

File: tests/gis_tests/inspectapp/tests.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def test_geom_columns(self):
         """
         out = StringIO()
         call_command('inspectdb',
```
Comment (timgraham) [inline tests/gis_tests/inspectapp/tests.py]: You might care to reformat these function calls to use the normal 4 space hanging indent style.

**Comment by timgraham** [comment]
Checkout errors on the precise build, but the checkouts that succeeded did pass, so I think this is good to go.

File: django/contrib/gis/db/backends/spatialite/features.py (hunk 1 above)
Comment (claudep) [inline django/contrib/gis/db/backends/spatialite/features.py]: Stupid tools... :-/

Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/features.py]: FWIW, I submitted several patches to isort and the author was very responsive. I just didn't care enough about this one to fix it.

**Pull request closed without merging**